src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,570 @@
+            }
+        }
+
+        if (parsed == null) {
+            // now try exact number patterns
+            parsed = new MessageFormat(patterns[TWO], locale).parseObject(source, parsePos);
+            if (parsed == null && !patterns[THREE].isEmpty()) {
+                parsed = new MessageFormat(patterns[THREE], locale).parseObject(source, parsePos);
+            }
+        }
+
+        // return the entire source if still no match
+        if (parsed == null) {
+            parsePos.setIndex(source.length());
+            parsed = new String[]{source};
+        }
+
+        if (parsed instanceof Object[] objs) {
+            return Arrays.asList(objs);
+        }
+        throw new InternalError("MessageFormat.parseObject() should return Object[]");
+    }
+
+    @Override
+    public AttributedCharacterIterator formatToCharacterIterator(Object arguments) {
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else {
+            throw new IllegalArgumentException("The arguments should be an Object list");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof ListFormat other) {
+            return locale.equals(other.locale) &&
+                Arrays.equals(patterns, other.patterns);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(locale, Arrays.hashCode(patterns));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
+            case 1 -> new MessageFormat("{0}", locale);
+            case 2, 3 -> {
+                var pattern = patterns[len + 1];
+                if (pattern != null && !pattern.isEmpty()) {
+                    yield new MessageFormat(pattern, locale);
+                } else {
+                    yield new MessageFormat(createMessageFormatString(len), locale);
+                }
+            }
+            default -> new MessageFormat(createMessageFormatString(len), locale);
+        };
+    }
+
+    private String createMessageFormatString(int count) {
+        var sb = new StringBuilder(patterns[START]);
```

Comment:
You might want to pre-size the StringBuilder to avoid too many resizes as strings are appended.

Reply:
Not sure what size would be optimal, so I simply gave 256 chars as an initial size